### PR TITLE
Material System Reports Warnings About PSO-Impacting Changes

### PR DESCRIPTION
**Pull request by santorac**
These changes make material system report warnings when gameplay scripts attempt to change PSO-impacting material properties at runtime. So far the material system has always allowed any properties to be changed at runtime, including those that affect Pipeline State Objects (PSOs), as this is supported on several platforms. But some platforms require that Pipeline State Objects be pre-compiled and shipped with the game. At some point we will need to add new restrictions that limit what material properties can be changed at runtime. In the meantime, these warnings should alert users to avoid this, as the functionality likely won't be supported in the future.

- Made the Material and LuaMaterialFunctor classes configurable to report errors or warnings when material properties modify Pipeline State Objects. This is controlled by a new "MaterialPropertyPsoHandling" enum.
- Made the EditorMaterialComponent override PSO handling as Enabled, to prevent warnings when the user is editing material instance property overrides. This requried a new MaterialComponentNotificationBus bus message "OnMaterialInstanceCreated".
- Added a new ScopedValue utility class that simply sets a value when it goes out of scope.
- Removed unnecessary GetMaterialPropertyDependencies member from material functor context classes, as this is already available as part of the functor itself.
- Made Material::SetPropertyValue return early when the property value hadn't actually changed. Besides being more efficientn, this prevents unnecessary spamming of the new warning.
- Made a couple imporvements to material_find_overrides_demo.lua: 
  - Made the target material slot name configurable through an exposed component property. 
  - Fixed a timing issue where the assignmentId was invalid if FindMaterialAssignmentId is called too early.

### CODE REVIEW

**Review by mrieggeramzn — APPROVED** [review]

File: Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#include <AzCore/std/functional.h>
+
+namespace AZ
+{
+    //! Sets a variable upon construction and again when the object goes out of scope.
+    template<typename T>
+    class ScopedValue
+    {
+    private:
+        T* m_ptr;
```
Comment (gadams3) [inline Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h:19]: Would this be safer as a reference? If not, should you assert that the pointer is not null in the constructor?

File: Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        T m_finalValue;
+
+    public:
+        ScopedValue(T* ptr, T initialValue, T finalValue)
+        {
+            m_ptr = ptr;
+            *m_ptr = initialValue;
+            m_finalValue = finalValue;
```
Comment (gadams3) [inline Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h:27]: The pointer and the final value can be initialized with the constructor

**Review by gadams3 — APPROVED** [review]

File: Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h (hunk 1 above)
Comment (santorac) [reply Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h:19]: Added an assert.

Comment (santorac) [reply Code/Framework/AtomCore/AtomCore/Utils/ScopedValue.h:19]: (I prefer pointers as the call site is more explicit in what is going on)

**Review by jeremyong-az — APPROVED** [review]
Thanks for this PR, it's a great step toward how PSO states could be managed in the future

**Pull request merged**